token/cli/src/main.rs

```diff
@@ -834,33 +779,52 @@ async fn command_transfer(
     no_wait: bool,
     allow_non_system_account_recipient: bool,
 ) -> CommandResult {
-    let mint_info = config.get_mint_info(&token, mint_decimals).await?;
+    let token = token_client_from_config(config, &token_pubkey);
+    let mint_info = config.get_mint_info(&token_pubkey, mint_decimals).await?;
+
+    // if the user got the decimals wrong, they may well have calculated the transfer amount wrong
+    if !config.sign_only && mint_decimals.is_some() && mint_decimals != Some(mint_info.decimals) {
+        return Err(format!(
+            "Decimals {} was provided, but actual value is {}",
+            mint_decimals.unwrap(),
+            mint_info.decimals
+        )
+        .into());
+    }
+
+    // decimals determines whether transfer_checked is used or not
+    let decimals = if use_unchecked_instruction {
+        None
+    } else if mint_decimals.is_some() {
+        mint_decimals
```

Comment:
the branch above is "if in online mode and decimals were provided, ensure it is correct." in offline mode, it defaults to 9 and could be incorrect. this three-way branch does unchecked if told to, uses the provided decimals if provided (online: verified correct but optional, offline: assumed correct but mandatory), and uses the mint decimals otherwise (online: correct by definition, offline: always 9, but this branch is never hit)

the behavior would be identical but maybe it would be clearer if i branched on `sign_only` instead

Reply:
actually i think ill just comment this better, checking `is_some()` makes it more robust to the callsite or cli interface getting refactored and passing `None` in in other cases